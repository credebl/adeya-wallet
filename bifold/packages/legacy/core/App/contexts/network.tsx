import { NetInfoStateType, useNetInfo } from '@react-native-community/netinfo'
import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import NetInfoModal from '../components/modals/NetInfoModal'
import { fetchLedgerNodes, canConnectToLedgerNode } from '../utils/ledger'

export interface NetworkContext {
  silentAssertConnectedNetwork: () => boolean
  assertConnectedNetwork: () => boolean
  displayNetInfoModal: () => void
  hideNetInfoModal: () => void
  assertLedgerConnectivity: () => Promise<boolean>
}

export const NetworkContext = createContext<NetworkContext>(null as unknown as NetworkContext)

export const NetworkProvider: React.FC = ({ children }) => {
  const netInfo = useNetInfo()
  const [isNetInfoModalDisplayed, setIsNetInfoModalDisplayed] = useState<boolean>(false)

  const displayNetInfoModal = () => {
    setIsNetInfoModalDisplayed(true)
  }

  const hideNetInfoModal = () => {
    setIsNetInfoModalDisplayed(false)
  }

  const silentAssertConnectedNetwork = () => {
    return netInfo.isConnected || netInfo.type !== NetInfoStateType.none
  }

  const assertConnectedNetwork = () => {
    const isConnected = silentAssertConnectedNetwork()
    if (!isConnected) {
      displayNetInfoModal()
    }

    return isConnected
  }

  const assertLedgerConnectivity = async (): Promise<boolean> => {
    const nodes = fetchLedgerNodes()

    if (typeof nodes === 'undefined' || nodes.length === 0) {
      return false
    }

    const connections = await Promise.all(nodes.map((n: { host: string; port: number }) => canConnectToLedgerNode(n)))
    return connections.includes(true)
  }

  return (
    <NetworkContext.Provider
      value={{
        silentAssertConnectedNetwork,
        assertConnectedNetwork,
        displayNetInfoModal,
        hideNetInfoModal,
        assertLedgerConnectivity,
      }}
    >
      {children}
      <NetInfoModal visible={isNetInfoModalDisplayed} onSubmit={() => hideNetInfoModal()} />
    </NetworkContext.Provider>
  )
}

export const useNetwork = () => useContext(NetworkContext)
