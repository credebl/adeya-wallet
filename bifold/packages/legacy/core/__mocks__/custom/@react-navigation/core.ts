const navigate = jest.fn()

const navigation = {
  navigate,
  setOptions: jest.fn(),
  getParent: jest.fn(() => ({
    navigate,
  })),
  getState: jest.fn(() => ({
    index: jest.fn(),
  })),
  goBack: jest.fn(),
  pop: jest.fn(),
  reset: jest.fn(),
}

const useNavigation = () => {
  return navigation
}

const useIsFocused = () => {
  return true
}

export { useNavigation, useIsFocused }
