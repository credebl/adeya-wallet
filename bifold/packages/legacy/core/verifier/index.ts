export { ProofRequestType } from './types/proof-reqeust-template'
export type {
  ProofRequestTemplate,
  DifProofRequestTemplatePayload,
  DifProofRequestTemplatePayloadData,
  IndyProofRequestTemplatePayload,
  IndyProofRequestTemplatePayloadData,
  IndyRequestedAttribute,
  IndyRequestedPredicate,
} from './types/proof-reqeust-template'
export type {
  GroupedSharedProofData,
  GroupedSharedProofDataItem,
  CredentialSharedProofData,
  ParsedIndyProof,
  UnresolvedPredicate,
  ResolvedPredicate,
  SharedAttributesGroup,
  SharedGroupedAttribute,
  SharedAttribute,
  MissingAttribute,
} from './types/proof'
export type { ProofCustomMetadata } from './types/metadata'
export { ProofMetadata } from './types/metadata'
export { useProofsByTemplateId } from './hooks/proofs'
export {
  getProofIdentifiers,
  parseIndyProof,
  groupSharedProofDataByCredential,
  getProofData,
  isPresentationReceived,
  isPresentationFailed,
  markProofAsViewed,
  linkProofWithTemplate,
} from './utils/proof'
export type { CreateProofRequestInvitationResult, SendProofRequestResult } from './utils/proof-request'
export {
  findProofRequestMessage,
  buildProofRequestDataForTemplate,
  createConnectionlessProofRequestInvitation,
  sendProofRequest,
  hasPredicates,
  isParameterizable,
} from './utils/proof-request'
export { defaultProofRequestTemplates } from './request-templates'
