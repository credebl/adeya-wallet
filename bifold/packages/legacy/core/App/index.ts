/* eslint-disable import/no-cycle */
import type { OnboardingStyleSheet } from './screens/Onboarding'

import { Agent } from '@aries-framework/core'
import AgentProvider from '@aries-framework/react-hooks'

import indyLedgers from '../configs/ledgers/indy'

import * as components from './components'
import Button, { ButtonType } from './components/buttons/Button'
import CheckBoxRow from './components/inputs/CheckBoxRow'
import CredentialCard from './components/misc/CredentialCard'
import InfoBox, { InfoBoxType } from './components/misc/InfoBox'
import ErrorModal from './components/modals/ErrorModal'
import NetInfo from './components/network/NetInfo'
import Record from './components/record/Record'
import InfoTextBox from './components/texts/InfoTextBox'
import { ToastType } from './components/toast/BaseToast'
import toastConfig from './components/toast/ToastConfig'
import { AttachTourStep } from './components/tour/AttachTourStep'
import { homeTourSteps } from './components/tour/HomeTourSteps'
import { TourBox } from './components/tour/TourBox'
import HomeContentView from './components/views/HomeContentView'
import * as contexts from './contexts'
import { AuthProvider } from './contexts/auth'
import { CommonUtilProvider } from './contexts/commons'
import { NetworkProvider } from './contexts/network'
import { useTour } from './contexts/tour/tour-context'
import { TourProvider } from './contexts/tour/tour-provider'
import { defaultConfiguration } from './defaultConfiguration'
import RootStack from './navigators/RootStack'
import AttemptLockout from './screens/AttemptLockout'
import Developer from './screens/Developer'
import OnboardingPages from './screens/OnboardingPages'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import UseBiometry from './screens/UseBiometry'
import * as types from './types'

export { LocalStorageKeys } from './constants'
export { initLanguages, initStoredLanguage, translationResources } from './localization'
export { ConfigurationProvider, useConfiguration } from './contexts/configuration'
export { defaultState, mergeReducers, StoreProvider, StoreContext, useStore } from './contexts/store'
export { default as Store, DispatchAction, reducer } from './contexts/reducers/store'
export { Assets as ImageAssets } from './theme'
export { ThemeProvider, useTheme } from './contexts/theme'
export { AnimatedComponentsProvider, useAnimatedComponents } from './contexts/animated-components'
export { ColorPallet } from './theme'
export { animatedComponents } from './animated-components'
export { theme } from './theme'
export { useAuth } from './contexts/auth'
export { NavigationTheme } from './theme'
export { testIdWithKey } from './utils/testable'
export { Screens, Stacks, TabStacks } from './types/navigators'
export { createStyles } from './screens/OnboardingPages'
export { statusBarStyleForColor, StatusBarStyles } from './utils/luminance'
export { BifoldError } from './types/error'
export { EventTypes } from './constants'

export type { AnimatedComponents } from './animated-components'
export type {
  ISVGAssets,
  IFontAttributes,
  IInputAttributes,
  IInputs,
  ITextTheme,
  IBrandColors,
  ISemanticColors,
  INotificationColors,
  IGrayscaleColors,
  IColorPallet,
  IAssets,
  ITheme,
} from './theme'
export type { ConfigurationContext } from './contexts/configuration'
export type { TourStep } from './contexts/tour/tour-context'
export type { GenericFn } from './types/fn'
export type { AuthenticateStackParams } from './types/navigators'
export type { OnboardingStyleSheet }
export type { WalletSecret } from './types/security'
export type { ReducerAction } from './contexts/reducers/store'
export type {
  State,
  Onboarding as OnboardingState,
  LoginAttempt as LoginAttemptState,
  Preferences as PreferencesState,
  Tours as ToursState,
} from './types/state'
export * from '../verifier'

export {
  indyLedgers,
  Agent,
  CommonUtilProvider,
  AgentProvider,
  AuthProvider,
  NetworkProvider,
  TourProvider,
  useTour,
  AttachTourStep,
  TourBox,
  homeTourSteps,
  Button,
  ButtonType,
  CheckBoxRow,
  CredentialCard,
  ErrorModal,
  InfoTextBox,
  InfoBox,
  InfoBoxType,
  ToastType,
  toastConfig,
  RootStack,
  NetInfo,
  OnboardingPages,
  Splash,
  Developer,
  Terms,
  HomeContentView,
  UseBiometry,
  AttemptLockout,
  Record,
  defaultConfiguration,
  types,
  components,
  contexts,
}
