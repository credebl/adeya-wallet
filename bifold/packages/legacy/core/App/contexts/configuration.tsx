import { IndyPoolConfig } from '@aries-framework/core'
import { createContext, ReducerAction, useContext } from 'react'

import { ProofRequestTemplate } from '../../verifier'
import { EmptyListProps } from '../components/misc/EmptyList'
import { RecordProps } from '../components/record/Record'
import OnboardingPages from '../screens/OnboardingPages'
import { ScanProps } from '../screens/Scan'
import { OCABundleResolver } from '../types/oca'
import { PINSecurityParams } from '../types/security'
import { SettingSection } from '../types/settings'

interface NotificationConfiguration {
  component: React.FC
  onCloseAction: (dispatch?: React.Dispatch<ReducerAction<any>>) => void
  title: string
  description: string
  buttonTitle: string
  pageTitle: string
}

export interface ConfigurationContext {
  pages: typeof OnboardingPages
  splash: React.FC
  terms: React.FC
  homeContentView: React.FC
  credentialListHeaderRight: React.FC
  credentialListOptions: React.FC
  credentialEmptyList: React.FC<EmptyListProps>
  developer: React.FC
  OCABundleResolver: OCABundleResolver
  scan: React.FC<ScanProps>
  useBiometry: React.FC
  record: React.FC<RecordProps>
  PINSecurity: PINSecurityParams
  indyLedgers: IndyPoolConfig[]
  settings: SettingSection[]
  customNotification: NotificationConfiguration
  useCustomNotifications: () => { total: number; notifications: any }
  connectionTimerDelay?: number
  autoRedirectConnectionToHome?: boolean
  proofRequestTemplates?: Array<ProofRequestTemplate>
  enableTours?: boolean
}

export const ConfigurationContext = createContext<ConfigurationContext>(null as unknown as ConfigurationContext)

export const ConfigurationProvider = ConfigurationContext.Provider

export const useConfiguration = () => useContext(ConfigurationContext)
