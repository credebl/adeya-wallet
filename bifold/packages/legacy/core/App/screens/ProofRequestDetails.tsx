import { useAgent } from '@aries-framework/react-hooks'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList, StyleSheet, Text, TextInput, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import {
  IndyProofRequestTemplatePayloadData,
  ProofRequestType,
  linkProofWithTemplate,
  sendProofRequest,
} from '../../verifier'
import Button, { ButtonType } from '../components/buttons/Button'
import AlertModal from '../components/modals/AlertModal'
import { useConfiguration } from '../contexts/configuration'
import { useTheme } from '../contexts/theme'
import { useTemplate } from '../hooks/proof-request-templates'
import { Screens, ProofRequestsStackParams } from '../types/navigators'
import { MetaOverlay, OverlayType } from '../types/oca'
import { Attribute, Field, Predicate } from '../types/record'
import { buildFieldsFromIndyProofRequestTemplate } from '../utils/oca'
import { parseSchemaFromId } from '../utils/schema'
import { testIdWithKey } from '../utils/testable'

type ProofRequestDetailsProps = StackScreenProps<ProofRequestsStackParams, Screens.ProofRequestDetails>

interface ProofRequestAttributesCardParams {
  data: IndyProofRequestTemplatePayloadData
  onChangeValue: (schema: string, label: string, name: string, value: string) => void
}

const AttributeItem: React.FC<{ item: Attribute }> = ({ item }) => {
  const { ListItems } = useTheme()

  const style = StyleSheet.create({
    attributeTitle: {
      ...ListItems.requestTemplateTitle,
      fontWeight: 'bold',
      fontSize: 18,
      paddingVertical: 8,
      marginRight: 8,
    },
  })

  return (
    <View style={{ flexDirection: 'row' }}>
      <Text style={style.attributeTitle}>{item.label || item.name}</Text>
      <Text style={style.attributeTitle}>{item.value}</Text>
    </View>
  )
}

const PredicateItem: React.FC<{
  item: Predicate
  onChangeValue: (name: string, value: string) => void
}> = ({ item, onChangeValue }) => {
  const { ListItems, ColorPallet } = useTheme()

  const style = StyleSheet.create({
    attributeTitle: {
      ...ListItems.requestTemplateTitle,
      fontWeight: 'bold',
      fontSize: 18,
      paddingVertical: 8,
      marginRight: 8,
    },
    input: {
      textAlign: 'center',
      borderBottomColor: ColorPallet.grayscale.black,
      borderBottomWidth: 1,
    },
  })

  return (
    <View style={{ flexDirection: 'row' }}>
      <Text style={style.attributeTitle}>{item.label || item.name}</Text>
      <Text style={style.attributeTitle}>{item.pType}</Text>
      {item.parameterizable && (
        <TextInput
          keyboardType="numeric"
          style={[style.attributeTitle, style.input]}
          onChangeText={(value) => onChangeValue(item.name || '', value)}
        >
          {item.pValue}
        </TextInput>
      )}
      {!item.parameterizable && <Text style={style.attributeTitle}>{item.pValue}</Text>}
    </View>
  )
}

const ProofRequestAttributesCard: React.FC<ProofRequestAttributesCardParams> = ({ data, onChangeValue }) => {
  const { ListItems, ColorPallet } = useTheme()
  const { i18n } = useTranslation()
  const { OCABundleResolver } = useConfiguration()

  const style = StyleSheet.create({
    credentialCard: {
      ...ListItems.requestTemplateBackground,
      padding: 16,
      borderRadius: 8,
      borderLeftWidth: 8,
      borderLeftColor: ColorPallet.brand.primary,
      marginBottom: 16,
    },
    schemaTitle: {
      ...ListItems.requestTemplateTitle,
      fontWeight: 'bold',
      fontSize: 20,
      paddingVertical: 8,
    },
    attributesList: {
      paddingLeft: 14,
    },
    attributesDelimiter: {
      width: '100%',
      height: 2,
      backgroundColor: ColorPallet.grayscale.lightGrey,
    },
  })

  const [meta, setMeta] = useState<MetaOverlay | undefined>(undefined)
  const [attributes, setAttributes] = useState<Field[] | undefined>(undefined)

  useEffect(() => {
    OCABundleResolver.resolve({ identifiers: { schemaId: data.schema }, language: i18n.language }).then((bundle) => {
      const metaOverlay = bundle?.metaOverlay || {
        captureBase: '',
        type: OverlayType.Meta10,
        name: parseSchemaFromId(data.schema).name,
        language: i18n.language,
      }
      setMeta(metaOverlay)
    })
  }, [data.schema])

  useEffect(() => {
    const attributes = buildFieldsFromIndyProofRequestTemplate(data)
    OCABundleResolver.presentationFields({
      identifiers: { schemaId: data.schema },
      attributes,
      language: i18n.language,
    }).then((fields) => {
      setAttributes(fields)
    })
  }, [data.schema])

  const countAttributes = attributes?.length

  const AttributeDelimiter: React.FC = () => <View style={style.attributesDelimiter} />

  return (
    <View style={[style.credentialCard]}>
      <Text style={style.schemaTitle}>{meta?.name}</Text>
      <FlatList
        style={style.attributesList}
        data={attributes}
        keyExtractor={(record, index) => record.name || index.toString()}
        renderItem={({ item, index }) => {
          return (
            <View>
              {item instanceof Attribute && <AttributeItem item={item as Attribute} />}
              {item instanceof Predicate && (
                <PredicateItem
                  item={item as Predicate}
                  onChangeValue={(name, value) => {
                    onChangeValue(data.schema, item.label || name, name, value)
                  }}
                />
              )}
              {index + 1 !== countAttributes && <AttributeDelimiter />}
            </View>
          )
        }}
      />
    </View>
  )
}

const ProofRequestDetails: React.FC<ProofRequestDetailsProps> = ({ route, navigation }) => {
  const { ColorPallet, TextTheme } = useTheme()
  const { t } = useTranslation()
  const { i18n } = useTranslation()
  const { OCABundleResolver } = useConfiguration()

  const { agent } = useAgent()
  if (!agent) {
    throw new Error('Unable to fetch agent from AFJ')
  }

  const style = StyleSheet.create({
    container: {
      flexGrow: 1,
      backgroundColor: ColorPallet.brand.primaryBackground,
      padding: 16,
    },
    header: {
      marginTop: 12,
      marginBottom: 36,
    },
    title: {
      color: TextTheme.title.color,
      fontSize: 28,
      fontWeight: 'bold',
    },
    description: {
      marginTop: 10,
      color: TextTheme.title.color,
      fontSize: 20,
    },
    footerButton: {
      marginTop: 'auto',
      marginBottom: 10,
    },
  })

  const { templateId, connectionId } = route?.params

  const [meta, setMeta] = useState<MetaOverlay | undefined>(undefined)
  const [attributes, setAttributes] = useState<Array<IndyProofRequestTemplatePayloadData> | undefined>(undefined)
  const [customPredicateValues, setCustomPredicateValues] = useState<Record<string, Record<string, number>>>({})
  const [invalidPredicate, setInvalidPredicate] = useState<
    { visible: boolean; predicate: string | undefined } | undefined
  >(undefined)

  const template = useTemplate(templateId)
  if (!template) {
    throw new Error('Unable to find proof request template')
  }

  useEffect(() => {
    const attributes = template.payload.type === ProofRequestType.Indy ? template.payload.data : []

    OCABundleResolver.resolve({ identifiers: { templateId }, language: i18n.language }).then((bundle) => {
      const metaOverlay = bundle?.metaOverlay || {
        captureBase: '',
        type: OverlayType.Meta10,
        name: template.name,
        description: template.description,
        language: i18n.language,
      }
      setMeta(metaOverlay)
      setAttributes(attributes)
    })
  }, [templateId])

  const onlyNumberRegex = /^\d+$/

  const onChangeValue = useCallback(
    (schema: string, label: string, name: string, value: string) => {
      if (!onlyNumberRegex.test(value)) {
        setInvalidPredicate({ visible: true, predicate: label })
        return
      }
      setInvalidPredicate(undefined)
      setCustomPredicateValues((prev) => ({
        ...prev,
        [schema]: {
          ...(prev[schema] || {}),
          [name]: parseInt(value),
        },
      }))
    },
    [setCustomPredicateValues, setInvalidPredicate]
  )

  const useProofRequest = useCallback(async () => {
    if (invalidPredicate) {
      setInvalidPredicate({ visible: true, predicate: invalidPredicate.predicate })
      return
    }

    if (connectionId) {
      // Send to specific contact and redirect to the chat with him
      sendProofRequest(agent, template, connectionId, customPredicateValues).then((result) => {
        if (result?.proofRecord) {
          linkProofWithTemplate(agent, result.proofRecord, templateId)
        }
      })

      navigation.getParent()?.navigate(Screens.Chat, { connectionId })
    } else {
      // Else redirect to the screen with connectionless request
      navigation.navigate(Screens.ProofRequesting, { templateId, predicateValues: customPredicateValues })
    }
  }, [agent, templateId, connectionId, customPredicateValues, invalidPredicate])

  const showTemplateUsageHistory = useCallback(async () => {
    navigation.navigate(Screens.ProofRequestUsageHistory, { templateId })
  }, [navigation, templateId])

  const Header: React.FC = () => {
    return (
      <View style={style.header}>
        <Text style={style.title}>{meta?.name}</Text>
        <Text style={style.description}>{meta?.description}</Text>
      </View>
    )
  }

  const Footer: React.FC = () => {
    return (
      <View>
        <View style={style.footerButton}>
          <Button
            title={connectionId ? t('Verifier.SendThisProofRequest') : t('Verifier.UseProofRequest')}
            accessibilityLabel={connectionId ? t('Verifier.SendThisProofRequest') : t('Verifier.UseProofRequest')}
            testID={connectionId ? testIdWithKey('SendThisProofRequest') : testIdWithKey('UseProofRequest')}
            buttonType={ButtonType.Primary}
            onPress={() => useProofRequest()}
          />
        </View>
        <View style={style.footerButton}>
          <Button
            title={t('Verifier.ShowTemplateUsageHistory')}
            accessibilityLabel={t('Verifier.ShowTemplateUsageHistory')}
            testID={testIdWithKey('ShowTemplateUsageHistory')}
            buttonType={ButtonType.Secondary}
            onPress={() => showTemplateUsageHistory()}
          />
        </View>
      </View>
    )
  }

  return (
    <SafeAreaView style={style.container} edges={['left', 'right']}>
      {invalidPredicate?.visible && (
        <AlertModal
          title={t('Verifier.InvalidPredicateValueTitle', { predicate: invalidPredicate.predicate })}
          message={t('Verifier.InvalidPredicateValueDetails')}
          submit={() => setInvalidPredicate({ visible: false, predicate: invalidPredicate.predicate })}
        />
      )}
      <FlatList
        ListHeaderComponent={Header}
        ListFooterComponent={Footer}
        style={{ backgroundColor: ColorPallet.brand.primaryBackground }}
        data={attributes}
        keyExtractor={(records) => records.schema}
        renderItem={({ item }) => <ProofRequestAttributesCard data={item} onChangeValue={onChangeValue} />}
        ListFooterComponentStyle={{ flexGrow: 1, justifyContent: 'flex-end' }}
        contentContainerStyle={{ flexGrow: 1 }}
      />
    </SafeAreaView>
  )
}

export default ProofRequestDetails
