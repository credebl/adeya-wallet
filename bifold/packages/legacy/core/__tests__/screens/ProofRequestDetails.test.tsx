import mockRNCNetInfo from '@react-native-community/netinfo/jest/netinfo-mock'
import { useNavigation } from '@react-navigation/core'
import { act, fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { ConfigurationContext } from '../../App/contexts/configuration'
import { NetworkProvider } from '../../App/contexts/network'
import configurationContext from '../contexts/configuration'
import { defaultProofRequestTemplates } from '../../verifier'
import ProofRequestDetails from '../../App/screens/ProofRequestDetails'
import { testIdWithKey } from '../../App'

jest.mock('react-native-permissions', () => require('react-native-permissions/mock'))
jest.mock('@react-native-community/netinfo', () => mockRNCNetInfo)
jest.mock('@react-navigation/core', () => {
  return require('../../__mocks__/custom/@react-navigation/core')
})
jest.mock('@react-navigation/native', () => {
  return require('../../__mocks__/custom/@react-navigation/native')
})
// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.mock('react-native-localize', () => {})
jest.mock('react-native-device-info', () => () => jest.fn())

jest.useFakeTimers('legacy')
jest.spyOn(global, 'setTimeout')

const templateId = defaultProofRequestTemplates[0].id
const connectionId = 'test'
const navigation = useNavigation()

describe('ProofRequestDetails Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  const renderView = (params: { templateId: string; connectionId?: string }) => {
    return render(
      <ConfigurationContext.Provider value={configurationContext}>
        <NetworkProvider>
          <ProofRequestDetails navigation={navigation as any} route={{ params: params } as any} />
        </NetworkProvider>
      </ConfigurationContext.Provider>
    )
  }

  test('Renders correctly', async () => {
    const tree = renderView({ templateId })
    await act(async () => null)
    expect(tree).toMatchSnapshot()
  })

  test('Schema and attributes are human readable', async () => {
    const tree = renderView({ templateId })

    const schema = await tree.findByText('Student full name', { exact: false })
    const credential = await tree.findAllByText('Student', { exact: false })
    const givenNames = await tree.findByText('First Name', { exact: false })
    const familyName = await tree.findByText('Last Name', { exact: false })

    expect(schema).not.toBe(null)
    expect(credential.length).toBe(3)
    expect(givenNames).not.toBe(null)
    expect(familyName).not.toBe(null)
  })

  test('Renders correctly when not pass connection id and pressing on a use proof request takes the user to a proof requesting screen', async () => {
    const tree = renderView({ templateId })

    await act(async () => {
      const useButton = tree.getByTestId(testIdWithKey('UseProofRequest'))

      fireEvent(useButton, 'press')

      expect(navigation.navigate).toBeCalledWith('Proof Requesting', {
        templateId,
        predicateValues: {},
      })
    })
  })

  test('Renders correctly when pass connection id and pressing on a send proof request takes the user to a chat', async () => {
    const tree = renderView({ templateId, connectionId })

    await act(async () => {
      const useButton = tree.getByTestId(testIdWithKey('SendThisProofRequest'))

      fireEvent(useButton, 'press')

      expect(navigation.navigate).toBeCalledWith('Chat', {
        connectionId,
      })
    })
  })
})
