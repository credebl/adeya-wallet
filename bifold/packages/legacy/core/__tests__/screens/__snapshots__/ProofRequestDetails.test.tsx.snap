// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProofRequestDetails Component Renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "left",
      "right",
    ]
  }
  style={
    Object {
      "backgroundColor": "#000000",
      "flexGrow": 1,
      "padding": 16,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    ListFooterComponentStyle={
      Object {
        "flexGrow": 1,
        "justifyContent": "flex-end",
      }
    }
    ListHeaderComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        Object {
          "requestedAttributes": Array [
            Object {
              "name": "student_first_name",
              "restrictions": Array [
                Object {
                  "cred_def_id": "XUxBrVSALWHLeycAUhrNr9:3:CL:26293:student_card",
                },
              ],
            },
            Object {
              "name": "student_last_name",
              "restrictions": Array [
                Object {
                  "cred_def_id": "XUxBrVSALWHLeycAUhrNr9:3:CL:26293:student_card",
                },
              ],
            },
          ],
          "schema": "XUxBrVSALWHLeycAUhrNr9:3:CL:26293:Student Card",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#000000",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "marginBottom": 36,
              "marginTop": 12,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 28,
                "fontWeight": "bold",
              }
            }
          >
            Student full name
          </Text>
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 20,
                "marginTop": 10,
              }
            }
          >
            Verify the full name of a student
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
                "borderLeftColor": "#42803E",
                "borderLeftWidth": 8,
                "borderRadius": 8,
                "marginBottom": 16,
                "padding": 16,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 20,
                "fontWeight": "bold",
                "paddingVertical": 8,
              }
            }
          >
            Student
          </Text>
          <View
            data={
              Array [
                Attribute {
                  "credentialId": undefined,
                  "encoding": undefined,
                  "format": undefined,
                  "label": "First Name",
                  "mimeType": undefined,
                  "name": "student_first_name",
                  "nonRevoked": undefined,
                  "restrictions": Array [
                    Object {
                      "cred_def_id": "XUxBrVSALWHLeycAUhrNr9:3:CL:26293:student_card",
                    },
                  ],
                  "revealed": undefined,
                  "revoked": undefined,
                  "type": "Text",
                  "value": null,
                },
                Attribute {
                  "credentialId": undefined,
                  "encoding": undefined,
                  "format": undefined,
                  "label": "Last Name",
                  "mimeType": undefined,
                  "name": "student_last_name",
                  "nonRevoked": undefined,
                  "restrictions": Array [
                    Object {
                      "cred_def_id": "XUxBrVSALWHLeycAUhrNr9:3:CL:26293:student_card",
                    },
                  ],
                  "revealed": undefined,
                  "revoked": undefined,
                  "type": "Text",
                  "value": null,
                },
              ]
            }
            getItem={[Function]}
            getItemCount={[Function]}
            keyExtractor={[Function]}
            onContentSizeChange={[Function]}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            style={
              Object {
                "paddingLeft": 14,
              }
            }
            viewabilityConfigCallbackPairs={Array []}
          >
            <View
              onLayout={[Function]}
              style={null}
            >
              <View>
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#000000",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "marginRight": 8,
                        "paddingVertical": 8,
                      }
                    }
                  >
                    First Name
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000000",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "marginRight": 8,
                        "paddingVertical": 8,
                      }
                    }
                  />
                </View>
                <View
                  style={
                    Object {
                      "backgroundColor": "#D3D3D3",
                      "height": 2,
                      "width": "100%",
                    }
                  }
                />
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View>
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#000000",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "marginRight": 8,
                        "paddingVertical": 8,
                      }
                    }
                  >
                    Last Name
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000000",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "marginRight": 8,
                        "paddingVertical": 8,
                      }
                    }
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "height": 0,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "marginBottom": 10,
                "marginTop": "auto",
              }
            }
          >
            <View
              accessibilityLabel="Verifier.UseProofRequest"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#42803E",
                  "borderRadius": 4,
                  "opacity": 1,
                  "padding": 16,
                }
              }
              testID="com.ariesbifold:id/UseProofRequest"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                      false,
                    ]
                  }
                >
                  Verifier.UseProofRequest
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 10,
                "marginTop": "auto",
              }
            }
          >
            <View
              accessibilityLabel="Verifier.ShowTemplateUsageHistory"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "borderColor": "#42803E",
                  "borderRadius": 4,
                  "borderWidth": 2,
                  "opacity": 1,
                  "padding": 16,
                }
              }
              testID="com.ariesbifold:id/ShowTemplateUsageHistory"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#42803E",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                      false,
                    ]
                  }
                >
                  Verifier.ShowTemplateUsageHistory
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
